capplets/time-admin/src/time-map.c

```diff
@@ -0,0 +1,629 @@
+                  gdouble pointx,
+                  gdouble pointy)
+{
+    static const double corner_radius = 9.0;
+    static const double margin_top = 12.0;
+    static const double margin_bottom = 12.0;
+    static const double margin_left = 24.0;
+    static const double margin_right = 24.0;
+
+    TimezoneMap *map = TIMEZONEMAP (widget);
+    GtkAllocation alloc;
+    PangoLayout *layout;
+    PangoRectangle text_rect;
+    double x;
+    double y;
+    double width;
+    double height;
+
+    if (!map->bubble_text)
+        return;
+
+    gtk_widget_get_allocation (widget, &alloc);
+    layout = gtk_widget_create_pango_layout (widget, NULL);
+    /* Layout the text */
+    pango_layout_set_alignment (layout, PANGO_ALIGN_CENTER);
+    pango_layout_set_spacing (layout, 3);
+    pango_layout_set_markup (layout, map->bubble_text, -1);
+
+    pango_layout_get_pixel_extents (layout, NULL, &text_rect);
+
+    /* Calculate the bubble size based on the text layout size */
+    width = text_rect.width + margin_left + margin_right;
+    height = text_rect.height + margin_top + margin_bottom;
+
+    if (pointx < alloc.width / 2)
+        x = pointx + 25;
+    else
+        x = pointx - width - 25;
+
+    y = pointy - height / 2;
+
+    /* Make sure it fits in the visible area */
+    x = CLAMP (x, 0, alloc.width - width);
+    y = CLAMP (y, 0, alloc.height - height);
+
+    cairo_save (cr);
+    cairo_translate (cr, x, y);
+
+    /* Draw the bubble */
+    cairo_new_sub_path (cr);
+    cairo_arc (cr, width - corner_radius, corner_radius, corner_radius, radians (-90), radians (0));
+    cairo_arc (cr, width - corner_radius, height - corner_radius, corner_radius, radians (0), radians (90));
+    cairo_arc (cr, corner_radius, height - corner_radius, corner_radius, radians (90), radians (180));
+    cairo_arc (cr, corner_radius, corner_radius, corner_radius, radians (180), radians (270));
+    cairo_close_path (cr);
+
+    cairo_set_source_rgba (cr, 0.2, 0.2, 0.2, 0.7);
+    cairo_fill (cr);
+
+    /* And finally draw the text */
+    cairo_set_source_rgb (cr, 1, 1, 1);
+    cairo_move_to (cr, margin_left, margin_top);
+    pango_cairo_show_layout (cr, layout);
+
+    g_object_unref (layout);
+    cairo_restore (cr);
+}
+
+static gboolean
+cc_timezone_map_draw (GtkWidget *widget,
+                      cairo_t   *cr)
+{
+    TimezoneMap *map = TIMEZONEMAP (widget);
+    g_autoptr(GdkPixbuf) orig_hilight = NULL;
+    GtkAllocation alloc;
+    g_autofree gchar *file = NULL;
+    g_autoptr(GError) err = NULL;
+    gdouble pointx, pointy;
+    char buf[16];
+
+    gtk_widget_get_allocation (widget, &alloc);
+
+    /* paint background */
+    gdk_cairo_set_source_pixbuf (cr, map->background, 0, 0);
+    cairo_paint (cr);
+
+    /* paint hilight */
+    if (gtk_widget_is_sensitive (widget))
+    {
+        file = g_strdup_printf ("/usr/share/mate-time-admin/map/timezone_%s.png",
+                                g_ascii_formatd (buf, sizeof (buf),
```

Comment:
There is a hard-coded path, maybe you can use pkgdatadir?

Reply:
Now use the definition in Makefile